### ISSUE TITLE
Disable arrow keys for selecting suggestions

### ISSUE THREAD

**Issue by tadashi-aikawa** [tadashi-aikawa/obsidian-various-complements-plugin#129]
@aubreyz 

> b) A bigger problem is the hijacking of the arrow keys. These are often used in editing, creating some disconcerting editing issues. One suggestion is that in the option "Additional cycle through suggestions keys" if one of these is selected then the usual arrow suggestion cycling is switched off. Or how about prefixing the suggestion with a number 1. 2. 3. etc -- then a suggestion can be chosen with Tab 1, Tab 2, Tab 3....

I see. What I can do is add a new option named "Disable arrow keys for selecting suggestions". What do you think?

_Originally posted by @tadashi-aikawa in https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/discussions/124#discussioncomment-2715361_

**Comment by Glint-Eye** [comment]
Yes. it would be very good to be able to disable arrow up/down keys especially if you work with lists a lot and jump around between items.

Also it would be great if "Right / Left Arrow" could be an option for cycling through suggestions as it is still very quick but not that disrupting. Usually I navigate inbetween words with ctrl + left / right anyways, so there isnt a conflict that often.

**Comment by tadashi-aikawa** [comment]
@aubreyz 
Added in [v7.0.0-beta8](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/discussions/133) 🚀

**Comment by aubreyz** [comment]
You are doing great work. 
On this one the arrows are disabled, but I think that it is not only a case of disabling them -- if the up or down arrow key is pressed it should exit the suggestion list and pass the arrow key to the main editor. That way it would be seamless.

**Comment by tadashi-aikawa** [comment]
Thanks for your confirming, @aubreyz :)

> if the up or down arrow key is pressed it should exit the suggestion list and pass the arrow key to the main editor

I implemented as you said, and confirmed it 🤔

**Comment by aubreyz** [comment]
Indeed. I think the overall emphasis (for seamless operation) should be that if the user presses any special navigation key other than the ones specifically marked for use to navigate the list or to select a word for completion, the list should close and the key should be passed to the editor. 

I think that is not just for the arrow case here but for everything

So if I type freq(downarrow), -- with arrows turned off -- a list should pop up and then disappear, but the overall effect should be as if the plugin was not installed. 

If exactly the same way if I currently type freq(forwardarrow) it works well

freq(Home) works well
freq(end) does not work **
freq(pageup) works well
freq(enter) works (presuming enter is not a selection key)
freq(tab) works (if tab is not a selection key) although the list remains open (may be a good thing?)

So my point is that it should be consistent - and at the moment it is except for arrows and end (of the ones I tested)

**Comment by tadashi-aikawa** [comment]
@aubreyz 
I'm sorry in words less than 😅 

In my environment, it works as you said (If I didn't misunderstand)

`fre(ArrowDown)` then `fre(ArrowUp)`

![2022-06-04](https://user-images.githubusercontent.com/9500018/171995819-4d356e79-aa36-41fa-b24f-5e369558077c.gif)

Is the above behavior not what you want? Or doesn't it work like the above in your environment?

**Comment by aubreyz** [comment]
> @aubreyz I'm sorry in words less than 😅
> 
> In my environment, it works as you said (If I didn't misunderstand)
> 
> `fre(ArrowDown)` then `fre(ArrowUp)`
> 
> ![2022-06-04](https://user-images.githubusercontent.com/9500018/171995819-4d356e79-aa36-41fa-b24f-5e369558077c.gif) [ ](https://user-images.githubusercontent.com/9500018/171995819-4d356e79-aa36-41fa-b24f-5e369558077c.gif) ![2022-06-04](https://user-images.githubusercontent.com/9500018/171995819-4d356e79-aa36-41fa-b24f-5e369558077c.gif) [ ](https://user-images.githubusercontent.com/9500018/171995819-4d356e79-aa36-41fa-b24f-5e369558077c.gif)
> 
> Is the above behavior not what you want? Or doesn't it work like the above in your environment?

No my environment does not do that. 
So it is a difference in behaviour (maybe combination of settings) not any misunderstanding. 

Your video shows exactly how it should be. When I press the up and down arrow what happens is just nothing -- i.e as if no key was pressed and the list remains open. 

Here is the first part of my data file:

```
{
  "strategy": "english-only",
  "matchStrategy": "prefix",
  "maxNumberOfSuggestions": 6,
  "maxNumberOfWordsAsPhrase": 3,
  "minNumberOfCharactersTriggered": 2,
  "minNumberOfWordsTriggeredPhrase": 1,
  "complementAutomatically": true,
  "delayMilliSeconds": 430,
  "disableSuggestionsDuringImeOn": false,
  "insertAfterCompletion": true,
  "firstCharactersDisableSuggestions": ":/^",
  "showMatchStrategy": true,
  "showComplementAutomatically": true,
  "showIndexingStatus": false,
  "descriptionOnSuggestion": "Short",
  "selectSuggestionKeys": "Shift+Space",
  "additionalCycleThroughSuggestionsKeys": "None",
  "disableUpDownKeysForCycleThroughSuggestionsKeys": true,
  "openSourceFileKey": "None",
  "propagateEsc": false,
  "enableCurrentFileComplement": false,
  "currentFileMinNumberOfCharacters": 0,
  "onlyComplementEnglishOnCurrentFileComplement": false,
  "enableCurrentVaultComplement": false,
  "currentVaultMinNumberOfCharacters": 0,
  "includeCurrentVaultPathPrefixPatterns": "",
  "excludeCurrentVaultPathPrefixPatterns": "",
  "includeCurrentVaultOnlyFilesUnderCurrentDirectory": false,
  "enableCustomDictionaryComplement": true,
  "customDictionaryPaths": "10 Utilities\\completions\\mydict.md\n10 Utilities\\completions\\mywords.md",
  "columnDelimiter": "Pipe",
  "customDictionaryWordRegexPattern": "",
  "delimiterToHideSuggestion": ";;;",
  "delimiterToDivideSuggestionsForDisplayFromInsertion": ">>>",
  "caretLocationSymbolAfterComplement": "<CARET>",
  "displayedTextSuffix": " => ...",
  "enableInternalLinkComplement": true,
  "suggestInternalLinkWithAlias": false,
  "excludeInternalLinkPathPrefixPatterns": "",
  "enableFrontMatterComplement": true,
  "frontMatterComplementMatchStrategy": "inherit",
  "insertCommaAfterFrontMatterCompletion": true,
  "showLogAboutPerformanceInConsole": false,
  "selectionHistoryTree": {
```

**Comment by tadashi-aikawa** [comment]
@aubreyz 
Thank you for sharing your settings, but it doesn't seem to be any problems.

What is your OS? (Windows / Mac / Android / iPhone / iPad...)

**Comment by aubreyz** [comment]
> What is your OS? (Windows / Mac / Android / iPhone / iPad...)

Windows 10

**Comment by aubreyz** [comment]
fol(right)
fol(left)
fol(up or down) -- this one does nothing and only way to exit is to click outside of the suggestion list

![arrows](https://user-images.githubusercontent.com/25100168/171998908-e9c708b3-3b00-4511-b943-d6b7cdaad070.gif)

**Comment by tadashi-aikawa** [comment]
Umm... I use Windows 10, too.

Could you tell me your settings by [this command](https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/3.%20Commands/%F0%9F%8E%AECopy%20plugin%20settings/)?

I want to adapt your settings directly to my environment.

**Comment by aubreyz** [comment]
{
    "version": "6.1.0",
    "mobile": false,
    "settings": {
        "strategy": "english-only",
        "matchStrategy": "prefix",
        "maxNumberOfSuggestions": 6,
        "maxNumberOfWordsAsPhrase": 3,
        "minNumberOfCharactersTriggered": 2,
        "minNumberOfWordsTriggeredPhrase": 1,
        "complementAutomatically": true,
        "delayMilliSeconds": 430,
        "disableSuggestionsDuringImeOn": false,
        "insertAfterCompletion": true,
        "firstCharactersDisableSuggestions": ":/^",
        "showMatchStrategy": true,
        "showComplementAutomatically": true,
        "showIndexingStatus": false,
        "descriptionOnSuggestion": "Short",
        "selectSuggestionKeys": "Shift+Space",
        "additionalCycleThroughSuggestionsKeys": "None",
        "disableUpDownKeysForCycleThroughSuggestionsKeys": true,
        "openSourceFileKey": "None",
        "propagateEsc": false,
        "enableCurrentFileComplement": false,
        "currentFileMinNumberOfCharacters": 0,
        "onlyComplementEnglishOnCurrentFileComplement": false,
        "enableCurrentVaultComplement": false,
        "currentVaultMinNumberOfCharacters": 0,
        "includeCurrentVaultPathPrefixPatterns": "",
        "excludeCurrentVaultPathPrefixPatterns": "",
        "includeCurrentVaultOnlyFilesUnderCurrentDirectory": false,
        "enableCustomDictionaryComplement": true,
        "customDictionaryPaths": "10 Utilities\\completions\\mydict.md\n10 Utilities\\completions\\mywords.md",
        "columnDelimiter": "Pipe",
        "customDictionaryWordRegexPattern": "",
        "delimiterToHideSuggestion": ";;;",
        "delimiterToDivideSuggestionsForDisplayFromInsertion": ">>>",
        "caretLocationSymbolAfterComplement": "<CARET>",
        "displayedTextSuffix": " => ...",
        "enableInternalLinkComplement": true,
        "suggestInternalLinkWithAlias": false,
        "excludeInternalLinkPathPrefixPatterns": "",
        "enableFrontMatterComplement": true,
        "frontMatterComplementMatchStrategy": "inherit",
        "insertCommaAfterFrontMatterCompletion": true,
        "showLogAboutPerformanceInConsole": false,
        "selectionHistoryTree": null
    }
}


Let me know what happens. Depending on that I could try in a fresh vault with no other plugins.

**Comment by aubreyz** [comment]
I ruled out a few things -- the problem is not altered by
a) Use of the up down arrows on the keyboard or the numpad
b) Choice of suggestion key
c) Choice of alternative up down keys in the list

**Comment by tadashi-aikawa** [comment]
I confirmed with a new vault and your settings but couldn't reproduce it.

1 Open `.obsidian\plugins\various-complements\data.json`
2. Replace with `"settings": ...` and remove `customDictionaryPaths: ...`
3. Restart Obsidian

Are there any errors on the devtools console in the Obsidian?

**Comment by aubreyz** [comment]
I created a new vault
I added the Complements plugin
I closed the vault and copied all the files from the current beta and my existing data.json to replace the installed plugin
I copied my custom dictionary across
Reopened Obsidian (is version 0.14.6)

Unfortunately the difficulty remains.....

I then deleted the data.json and restarted obsidian, and then changed only two default settings 
a) Custom Dictionary on (which gets your default online one)
b) Disable the up/down arrows for selection

Unfortunately still your video is not replicated......

So am not sure
a) Is it at all possible that you are using a Beta even later than the 7.0.0-beta8 on Github?
b) Some sort of difference in the keyboard drivers for different countries that makes it detect arrow keys from your code differently???

**Comment by tadashi-aikawa** [comment]
> a) Is it at all possible that you are using a Beta even later than the 7.0.0-beta8 on Github?

If you installed with BRAT, The version is 7.0.0-beta8. The below is my settings (new vault).

```json
{
    "version": "7.0.0-beta8",
    "mobile": false,
    "settings": {
        ....
    }
}
```

Could you confirm the version in your settings by [this command](https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/3.%20Commands/%F0%9F%8E%AECopy%20plugin%20settings/)?

> b) Some sort of difference in the keyboard drivers for different countries that makes it detect arrow keys from your code

I confirm the keycode of `ArrowUp/ArrowDown` from https://www.toptal.com/developers/keycode.

![image](https://user-images.githubusercontent.com/9500018/172013214-313094b1-b886-408e-b69e-2bde6ceb5963.png)

Could you tell me the result?

**Comment by aubreyz** [comment]
Ahh.. The way I installed the beta is to simply overwrite all of the existing

styles.css
manifest.json
manifest-beta.json
main.js

From the beta, while keeping the dat file. 

All the new beta features appeared on the menu, but the version is still reported as 6.1.0 on the dump

However the manifest.json in the beta for version 7 downloaded here
https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/releases/tag/7.0.0-beta8
Says 6.1.0 in the code

The [manifest-beta.json](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/releases/download/7.0.0-beta8/manifest-beta.json)
says  "version": "7.0.0-beta8",

But obviously the code in main.json [main.js](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/releases/download/7.0.0-beta8/main.js) has just one version. 

If I replace the manifest.json with manifest-beta.json the version is reported differently but nothing else changes

**Comment by aubreyz** [comment]
The downarrow key is reported as 40 with the same other settings as yours

**Comment by aubreyz** [comment]
Apart from knowing how to get into the devtools (Control Shift I), I have no experience to know where I would look for errors within that.

**Comment by tadashi-aikawa** [comment]
You can see error logs as follows.

![image](https://user-images.githubusercontent.com/9500018/172032846-65caf1fd-3bf2-457f-9524-bddd94302c4a.png)

**Comment by tadashi-aikawa** [comment]
I could reproduce it in the following case.

```
dev
ge
```

![2022-06-05](https://user-images.githubusercontent.com/9500018/172033094-b002531a-68e9-42d7-b634-a102477e05be.gif)

**Comment by tadashi-aikawa** [comment]
@aubreyz 
Released `v7.0.0-beta9` 🚀 

I'm not confident, but please try it.

**Comment by aubreyz** [comment]
The beta9 solves it!!!!
So something between the beta8 and the beta9

**Comment by tadashi-aikawa** [comment]
Greaaaaaat!! Thanks a lot 😄

**Comment by tadashi-aikawa** [comment]
Released in [v7.0.0](https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/releases/tag/7.0.0) 🚀
